sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala

```diff
@@ -32,23 +32,34 @@ object DistributionAndOrderingUtils {
     case write: RequiresDistributionAndOrdering =>
       val resolver = conf.resolver
 
-      val distribution = write.requiredDistribution match {
+      val numParts = write.requiredNumPartitions()
+      val (distribution, numPartitions) = write.requiredDistribution match {
         case d: OrderedDistribution =>
-          d.ordering.map(e => toCatalyst(e, query, resolver))
+          val dist = d.ordering.map(e => toCatalyst(e, query, resolver))
+          (dist, numParts)
         case d: ClusteredDistribution =>
-          d.clustering.map(e => toCatalyst(e, query, resolver))
+          val dist = d.clustering.map(e => toCatalyst(e, query, resolver))
+          (dist, numParts)
         case _: UnspecifiedDistribution =>
-          Array.empty[Expression]
+          (Array.empty[Expression], numParts)
       }
 
       val queryWithDistribution = if (distribution.nonEmpty) {
-        val numShufflePartitions = conf.numShufflePartitions
+        val finalNumPartitions = if (numPartitions > 0) {
+          numPartitions
+        } else {
+          conf.numShufflePartitions
+        }
         // the conversion to catalyst expressions above produces SortOrder expressions
         // for OrderedDistribution and generic expressions for ClusteredDistribution
         // this allows RepartitionByExpression to pick either range or hash partitioning
-        RepartitionByExpression(distribution, query, numShufflePartitions)
+        RepartitionByExpression(distribution, query, finalNumPartitions)
       } else {
-        query
+        if (numPartitions > 0) {
+          Repartition(numPartitions, shuffle = true, query)
```

Comment:
@aokolnychyi, what do you think about using a round-robin partition when there is no required distribution?

Reply:
@aokolnychyi kindly reminder.